doc/l1sign.1
.\" l1sign - Implementation of the Lamport-Diffie one-time signature scheme
.\" Copyright (c) 2019  Janik Rabe <info@janikrabe.com>
.\"
.\" This program is free software: you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with this program.  If not, see <https://www.gnu.org/licenses/>.

.TH l1sign 1 "2020-04-16" "version 0.2.0"

.SH NAME

\fBl1sign\fP - Implementation of the Lamport-Diffie one-time signature scheme

.SH SYNOPSIS

\fBl1sign\fP [\fIOPTIONS\fP] <\fICOMMAND\fP> [\fIARGUMENTS\fP]

.SH DESCRIPTION

\fBl1sign\fP is an implementation of the Lamport-Diffie one-time signature
scheme.
It allows users to create key pairs consisting of a secret key and a public
key.
The secret key can be used to sign a message at a later time.
Anyone with a copy of a public key, a message, and a corresponding signature
can verify the validity of the signature.

The security of the Lamport-Diffie signature scheme depends on the randomness
of the secret key and the security and output size of a cryptographic hash
function.
A secret key must only be used to sign a single message; each additional
signature reduces the security of the scheme by approximately one half.

\fBl1sign\fP allows users to generate secret and public keys, sign messages,
and verify message signatures.

If "-" is specified instead of a file name, \fBl1sign\fP uses standard input or
standard output, depending on whether the file is read from or written to.

.SH OPTIONS

\fB\-H, \-\-hash\fP=\fINAME\fP
.RS 4
Use the specified hash function.
Supported hash functions include
\fBblake2b_160\fP,
\fBblake2b_256\fP,
\fBblake2b_384\fP,
\fBblake2b_512\fP,
\fBblake2s_128\fP,
\fBblake2s_160\fP,
\fBblake2s_224\fP,
\fBblake2s_256\fP,
\fBgostr3411_94\fP,
\fBsha3-224\fP,
\fBsha3-256\fP,
\fBsha3-384\fP,
\fBsha3-512\fP,
\fBshake128\fP,
\fBshake256\fP,
\fBsha224\fP,
\fBsha256\fP,
\fBsha384\fP,
\fBsha512\fP,
\fBstribog256\fP,
\fBstribog512\fP, and
\fBwhirlpool\fP.
Note that this argument allows users to specify a number of non-cryptographic
hash functions.  It is the user's responsibility to ensure that the specified
hash function is secure.
By default, \fBblake2b_512\fP is used.
.RE

\fB\-m, \-\-message\fP=\fIFILE\fP
.RS 4
Specify the file to be signed or verified.
.RE

\fB\-v, \-\-verbose\fP
.RS 4
Print diagnostic information during the operation.
.RE

\fB\-h, \-\-help\fP
.RS 4
Print the help and exit.
.RE

.SH COMMANDS

\fBgenkey\fP <\fIsecret-key.l1sec\fP>
.RS 4
Generate a random secret key and save it to \fIsecret-key.l1sec\fP.
If the output file already exists, it is overwritten.
.RE

\fBpubkey\fP <\fIsecret-key.l1sec\fP> <\fIpublic-key.l1pub\fP>
.RS 4
Generate the public key corresponding to secret key \fIsecret-key.l1sec\fP and
save it to \fIpublic-key.l1pub\fP.
If the output file already exists, it is overwritten.
.RE

\fBsign\fP <\fIsecret-key.l1sec\fP> <\fIsignature.l1sig\fP>
.RS 4
Sign the message given by the \fB\-\-message\fP option with secret key
\fIsecret-key.l1sec\fP and save the resulting signature to
\fIsignature.l1sig\fP.
If the output file already exists, it is overwritten.
.RE

\fBverify\fP <\fIpublic-key.l1pub\fP> <\fIsignature.l1sig\fP>
.RS 4
Check whether \fIsignature.l1sig\fP is a valid signature for the message given
by the \fB\-\-message\fP option and was generated with the secret key
corresponding to the public key \fIpublic-key.l1pub\fP.
.RE

.SH EXAMPLES

Generate a random secret key:
.RS 4
.Bd
\fBl1sign\fP genkey \fIexample.l1sec\fP
.Ed
.RE

Generate the corresponding public key:
.RS 4
.Bd
\fBl1sign\fP pubkey \fIexample.l1sec\fP \fIexample.l1pub\fP
.Ed
.RE

Sign a message:
.RS 4
.Bd
\fBl1sign\fP -m \fImessage.txt\fP sign \fIexample.l1sec\fP \fIexample.l1sig\fP
.Ed
.RE

Verify a signature:
.RS 4
.Bd
\fBl1sign\fP -m \fImessage.txt\fP verify \fIexample.l1pub\fP \fIexample.l1sig\fP
.Ed
.RE

.SH SECURITY

\fBl1sign\fP has not received an independent security audit.
We recommend that you use this program only in conjunction with an alternative
implementation or signature scheme.

\fBl1sign\fP does not delete secret keys after they are used to create a
signature.
It is the user's responsibility to ensure that each key is used only once.

By default, \fBl1sign\fP stores sensitive information such as secret keys in
secure memory pages that cannot be swapped out.
However, some features such as hibernation (or "suspend to disk"), if used
while \fBl1sign\fP is running, may nevertheless result in sensitive information
being written to non-volatile storage, from where it may be recoverable later.

.SH AUTHOR

Janik Rabe <info@janikrabe.com>
.br
.RS 4
https://janikrabe.com
.PP


NEWS
Changes in version 0.2.0

- Imported documentation from website.

Changes in version 0.1.0

- Initial release.


src/l1sign.c
/*
 * l1sign - Implementation of the Lamport-Diffie one-time signature scheme
 * Copyright (c) 2019  Janik Rabe <info@janikrabe.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

#include "l1sign.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "l1sign_gcrypt.h"

#include "l1sign_cmd_genkey.h"
#include "l1sign_cmd_pubkey.h"
#include "l1sign_cmd_sign.h"
#include "l1sign_cmd_verify.h"

#include <config.h>

static const struct command commands[] = {
	{
		"genkey",
		"Generate a random private key",
		l1_cmd_genkey,
	},
	{
		"pubkey",
		"Generate a public key from a private key",
		l1_cmd_pubkey,
	},
	{
		"sign",
		"Sign a message with a private key",
		l1_cmd_sign,
	},
	{
		"verify",
		"Verify a message signature",
		l1_cmd_verify,
	},
	{
		NULL,
		NULL,
		NULL,
	},
};

const struct command *find_command(const char *name) {
	for (size_t i = 0; commands[i].name; ++i) {
		if (!strcmp(name, commands[i].name)) {
			return &commands[i];
		}
	}

	return NULL;
}

void print_header(void) {
	printf("%s by %s <%s>\n", PACKAGE_STRING,
			PACKAGE_AUTHOR, PACKAGE_BUGREPORT);
	printf("%s\n\n", PACKAGE_URL);
}

void print_cmd_usage(char *usage) {
	print_header();
	fprintf(stderr, "Usage: " PACKAGE_NAME " [options] %s\n", usage);
}

void print_usage(FILE *out) {
	char usage[] =
			"Usage: " PACKAGE_NAME " [options] <command> [args]\n\n"
			"Commands:\n";
	fputs(usage, out);

	for (size_t i = 0; commands[i].name; ++i) {
		fprintf(out,
				"  %s: %*s%s\n",
				commands[i].name,
				8 - (int) strlen(commands[i].name), "",
				commands[i].description);
	}
}

void print_arg_required(char *opt) {
	fprintf(stderr, "Option '%s' requires an argument\n", opt);
}

void print_opt_accept(char *cmd, char *opt) {
	fprintf(stderr, "Command '%s' requires option '--%s'\n", cmd, opt);
}

void print_opt_reject(char *cmd, char *opt) {
	fprintf(stderr, "Command '%s' does not accept option '--%s'\n", cmd, opt);
}

int main(int argc, char **argv) {
	const struct command *cmd;
	struct options opts = { 0 };
	int next = 0;

	while (argv[++next] && argv[next][0] == '-') {
		if (!strcmp(argv[next], "-H") || !strcmp(argv[next], "--hash")) {
			char *hash_name = argv[++next];

			if (!hash_name) {
				print_arg_required(argv[next - 1]);
				return EXIT_FAILURE;
			}

			if (!(opts.hash = gcry_md_map_name(hash_name))) {
				fprintf(stderr, "Unknown hash algorithm: %s\n", hash_name);
				return EXIT_FAILURE;
			}
		} else if (!strcmp(argv[next], "-m") || !strcmp(argv[next], "--message")) {
			opts.message = argv[++next];

			if (!opts.message) {
				print_arg_required(argv[next - 1]);
				return EXIT_FAILURE;
			}
		} else if (!strcmp(argv[next], "-v") || !strcmp(argv[next], "--verbose")) {
			opts.verbose = true;
		} else if (!strcmp(argv[next], "-h") || !strcmp(argv[next], "--help")) {
			print_header();
			print_usage(stdout);
			return EXIT_SUCCESS;
		} else if (!strcmp(argv[next], "--")) {
			++next;
			break;
		} else {
			fprintf(stderr, "Unknown option: %s\n", argv[next]);
			return EXIT_FAILURE;
		}
	}

	if (!argv[next]) {
		print_header();
		print_usage(stdout);
		return EXIT_SUCCESS;
	}

	cmd = find_command(argv[next]);

	if (!cmd) {
		fprintf(stderr, "No such command: %s\n", argv[next]);
		print_usage(stderr);
		return EXIT_FAILURE;
	}

	++next;

	if (!opts.hash) {
		opts.hash = GCRY_MD_BLAKE2B_512;
	}

	if (opts.verbose) {
		unsigned int hash_bytes = l1_gcry_hash_nbytes(opts.hash);
		fprintf(stderr, "Hash: %s (%d bits)\n",
				gcry_md_algo_name(opts.hash),
				hash_bytes * 8);
	}

	if (!l1_gcry_init(opts.hash)) {
		return EXIT_FAILURE;
	}

	if (atexit(l1_gcry_term)) {
		fprintf(stderr, "Failed to register exit handler\n");
		l1_gcry_term();
		return EXIT_FAILURE;
	}

	return cmd->invoke(&opts, argc - next, &argv[next]);
}
